packages/@uppy/box/src/Box.tsx

```diff
@@ -1,30 +1,66 @@
-import { UIPlugin } from '@uppy/core'
-import { Provider, getAllowedHosts, tokenStorage } from '@uppy/companion-client'
+import {
+  Provider,
+  getAllowedHosts,
+  tokenStorage,
+  type CompanionPluginOptions,
+} from '@uppy/companion-client'
+import { UIPlugin, Uppy } from '@uppy/core'
 import { ProviderViews } from '@uppy/provider-views'
-import { h } from 'preact'
+import { h, type ComponentChild } from 'preact'
 
-import locale from './locale.js'
+import type { UppyFile, Body, Meta } from '@uppy/utils/lib/UppyFile'
+import type { UnknownProviderPluginState } from '@uppy/core/lib/Uppy.ts'
+import locale from './locale.ts'
+// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+// @ts-ignore We don't want TS to generate types for the package.json
 import packageJson from '../package.json'
 
-export default class Box extends UIPlugin {
+export type BoxOptions = CompanionPluginOptions
+
+export default class Box<M extends Meta, B extends Body> extends UIPlugin<
+  BoxOptions,
+  M,
+  B,
+  UnknownProviderPluginState
+> {
   static VERSION = packageJson.version
 
-  constructor (uppy, opts) {
+  icon: () => JSX.Element
```

Comment:
I thought `ComponentChild` was preferable over `JSX.Element`, or is there an exception in this case?
```suggestion
  icon: () => ComponentChild
```